### PR TITLE
YUMMY-006 Create upload photo flow

### PR DESCRIPTION
**Pull request by htorrex**
add a new meal functionality and using text wrapper layout
Adjusting the service call to use it properly
Adding + button, adding toolbar with coordinator layout
Updating to Retrofit beta2.

### CODE REVIEW

File: app/src/main/java/com/visiontech/yummysmile/ui/activity/CreateMealActivity.java [hunk 1]
```diff
@@ -0,0 +1,368 @@
...
+public class CreateMealActivity extends AppCompatActivity implements CreateMealView {
+
+    private static final String LOG_TAG = CreateMealActivity.class.getName();
+    private static final int REQUEST_IMAGE_CAPTURE = 100;
+    private static final int REQUEST_WRITE_EXTERNAL_STORAGE = 200;
+    private String currentPhotoPath;
+    private CreateMealPresenter createMealPresenter;
+
+    //UI
+    private EditText mealName;
```
Comment (manuelortiz28) [inline app/src/main/java/com/visiontech/yummysmile/ui/activity/CreateMealActivity.java]: Is a good practice to have attributes ordered by accesor and data type

File: app/src/main/java/com/visiontech/yummysmile/ui/activity/CreateMealActivity.java [hunk 2]
```diff
@@ -0,0 +1,368 @@
...
+                requestPermissions();
+            }
+        });
+    }
+
+    @Override
+    protected void onResume() {
+        Log.d(LOG_TAG, "onResume()");
+        super.onResume();
+        createMealPresenter = new CreateMealPresenter(this, CreateMealActivity.this);
```
Comment (manuelortiz28) [inline app/src/main/java/com/visiontech/yummysmile/ui/activity/CreateMealActivity.java]: Why don't put this code inside the onCreate method?

File: app/src/main/java/com/visiontech/yummysmile/ui/activity/CreateMealActivity.java [hunk 3]
```diff
@@ -0,0 +1,368 @@
...
+        }
+    }
+
+    /**
+     * Method that create the file to save on disk.
+     */
+    private File createImageFile() throws IOException {
+        Log.d(LOG_TAG, "4_createImageFile()_");
+        // Create an image file name
+        String timeStamp = new SimpleDateFormat("yyyyMMdd_HHmmss").format(new Date());
```
Comment (manuelortiz28) [inline app/src/main/java/com/visiontech/yummysmile/ui/activity/CreateMealActivity.java]: You could put a constant for the string. And maybe we could create a DateUtils class with format methods

File: app/src/main/java/com/visiontech/yummysmile/ui/activity/CreateMealActivity.java [hunk 4]
```diff
@@ -0,0 +1,368 @@
...
+
+    /**
+     * Method that read the image from the device and then show it on the device.
+     */
+    private void setPicture() {
+        Log.d(LOG_TAG, "6_setPicture()_");
+        // Get the dimensions of the View
+//        int targetW = mealPicture.getWidth();
+        //Fixme figure out the best way to do it.
+        int targetW = 400;
```
Comment (manuelortiz28) [inline app/src/main/java/com/visiontech/yummysmile/ui/activity/CreateMealActivity.java]: You could use dimens file

File: app/src/main/java/com/visiontech/yummysmile/ui/activity/CreateMealActivity.java [hunk 5]
```diff
@@ -0,0 +1,368 @@
...
+
+            //Getting full size image
+            setPicture();
+        }
+    }
+
+    /**
+     * Method that read the image from the device and then show it on the device.
+     */
+    private void setPicture() {
```
Comment (manuelortiz28) [inline app/src/main/java/com/visiontech/yummysmile/ui/activity/CreateMealActivity.java]: We could split this method into other more generic. For example create a new method that receives the file path and returns a bitmap.

File: app/src/main/java/com/visiontech/yummysmile/ui/activity/CreateMealActivity.java [hunk 6]
```diff
@@ -0,0 +1,368 @@
...
+            Log.d(LOG_TAG, "_IOException_Message" + e.getMessage());
+            Log.d(LOG_TAG, "_IOException_Cause" + e.getCause());
+            e.printStackTrace();
+        }
+
+        mealPicture.setImageBitmap(bitmap);
+        mealPicture.setTag(currentPhotoPath);
+    }
+
+    private void hideKeyboard() {
```
Comment (manuelortiz28) [inline app/src/main/java/com/visiontech/yummysmile/ui/activity/CreateMealActivity.java]: This method could be in an Util class

File: app/src/main/java/com/visiontech/yummysmile/ui/controller/MealsControllerImpl.java [hunk 7]
```diff
@@ -57,4 +62,39 @@ public void onNext(MealsDTO mealsDTO) {
...
+
+    @Override
+    public void createMeal(JsonObject meal, File photo) {
+        Log.d(LOG_TAG, "createMeal()");
+
+        // Create an instance of our API interface.
+        MealAPIService mealAPIService = FactoryRestAdapter.createRetrofitService(MealAPIService.class);
+
+        // Creating the request body base on the file
+        RequestBody requestBody = RequestBody.create(MediaType.parse("multipart/form-data"), photo);
```
Comment (manuelortiz28) [inline app/src/main/java/com/visiontech/yummysmile/ui/controller/MealsControllerImpl.java]: You could use a Constants file for the string

File: app/src/main/java/com/visiontech/yummysmile/ui/presenter/CreateMealPresenter.java [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+        JsonObject jsonObject = new JsonObject();
+        jsonObject.addProperty("name", name);
+        mealsController.createMeal(jsonObject, photo);
+    }
+
+    @Override
+    public <T> void onNext(T data) {
+        Log.d(LOG_TAG, "onNext()");
+        JsonObject jsonResponse = (JsonObject) data;
+        if (jsonResponse.has("objectId") && jsonResponse.get("objectId") != null) {
```
Comment (manuelortiz28) [inline app/src/main/java/com/visiontech/yummysmile/ui/presenter/CreateMealPresenter.java]: Please use constants

File: app/src/main/java/com/visiontech/yummysmile/ui/presenter/CreateMealPresenter.java [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+            String objectId = jsonResponse.get("objectId").getAsString();
+            Log.d(LOG_TAG, "objectId: " + objectId);
+        }
+
+        if (jsonResponse.has("filename") && jsonResponse.get("filename") != null) {
```
Comment (manuelortiz28) [inline app/src/main/java/com/visiontech/yummysmile/ui/presenter/CreateMealPresenter.java]: Please use constants

File: app/src/main/res/layout/create_meal.xml [hunk 10]
```diff
@@ -0,0 +1,121 @@
...
+            android:text="@string/meal_title"
+            />
+
+        <android.support.design.widget.TextInputLayout
+            android:id="@+id/ti_name_wrapper"
+            android:layout_width="match_parent"
+            android:layout_height="wrap_content"
+            android:layout_below="@+id/tv_title"
+            app:counterEnabled="true"
+            app:counterMaxLength="30"
```
Comment (manuelortiz28) [inline app/src/main/res/layout/create_meal.xml]: You could use an integer value in the xml file

File: app/src/main/res/layout/create_meal.xml [hunk 11]
```diff
@@ -0,0 +1,121 @@
...
+            app:errorEnabled="true">
+
+            <EditText
+                android:id="@+id/et_name"
+                style="@style/Font.Roboto.Medium.Body2"
+                android:layout_width="match_parent"
+                android:layout_height="wrap_content"
+                android:hint="@string/meal_name_hint"
+                android:inputType="text"
+                android:maxLength="30"
```
Comment (manuelortiz28) [inline app/src/main/res/layout/create_meal.xml]: You could use an integer value in the xml file

File: app/src/main/res/layout/create_meal.xml [hunk 12]
```diff
@@ -0,0 +1,121 @@
...
+                />
+        </android.support.design.widget.TextInputLayout>
+
+        <android.support.design.widget.TextInputLayout
+            android:id="@+id/ti_description_wrapper"
+            android:layout_width="match_parent"
+            android:layout_height="wrap_content"
+            android:layout_below="@+id/ti_name_wrapper"
+            app:counterEnabled="true"
+            app:counterMaxLength="100">
```
Comment (manuelortiz28) [inline app/src/main/res/layout/create_meal.xml]: You could use an integer value in the xml file

File: app/src/main/res/layout/create_meal.xml [hunk 13]
```diff
@@ -0,0 +1,121 @@
...
+
+            <EditText
+                android:id="@+id/et_description"
+                style="@style/Font.Roboto.Medium.Body2"
+                android:layout_width="match_parent"
+                android:layout_height="wrap_content"
+                android:hint="@string/meal_description_hint"
+                android:inputType="textMultiLine"
+                android:maxLength="100"
```
Comment (manuelortiz28) [inline app/src/main/res/layout/create_meal.xml]: You could use an integer value in the xml file

File: app/src/main/res/layout/create_meal.xml [hunk 14]
```diff
@@ -0,0 +1,121 @@
...
+            android:paddingTop="@dimen/activity_vertical_margin">
+
+            <TextView
+                android:id="@+id/tv_title_img_source"
+                style="@style/Font.Roboto.Regular.Caption"
+                android:layout_width="wrap_content"
+                android:layout_height="wrap_content"
+                android:layout_below="@+id/ti_description_wrapper"
+                android:gravity="left"
+                android:paddingRight="5dp"
```
Comment (manuelortiz28) [inline app/src/main/res/layout/create_meal.xml]: Please use dimens file

File: app/src/main/res/layout/create_meal.xml [hunk 15]
```diff
@@ -0,0 +1,121 @@
...
+                android:text="@string/meal_source_picture"
+                android:visibility="gone"
+                />
+
+            <ImageView
+                android:id="@+id/iv_meal_picture"
+                android:layout_width="100dp"
```
Comment (manuelortiz28) [inline app/src/main/res/layout/create_meal.xml]: Please use dimens file

**Comment by manuelortiz28** [comment]
@htorrex Are you gonna update this PR? If not, you could close it

**Comment by htorrex** [comment]
Closing this one since is not longer valid we refer to: https://github.com/manuelortiz28/yummysmile/pull/16

**Pull request closed without merging**